build: define FORCE_VEX_ENCODING=1 for nasm if compilers have avx

Andarwinux opened the pull request:
use a x86inc.asm feature to automatically convert sse asm to avx when compiling for avx targets

this fixes an edge issue in a special case where libass is built with -mavx but not -mavx2, and vzeroupper has been disabled. In addition, since DCE currently doesn't work for asm, this also reduces the number of sse instructions in the final build product targeted for avx2+.

Andarwinux commented:
FYI, with this change, libass built with -march=native no longer includes any SSE instructions.

<details><summary>bin-cpuflags-x86 output</summary>
<p>

Format: ELF
Architecture: x86-64

Instructions
------------
Note: instructions that belong to multiple feature sets make counters overlap.
= 55602

X64 25860 (46.51%)
    mov    7732 (13.91%)
    je     1929 (3.47%)
    lea    1927 (3.47%)
    add    1529 (2.75%)
    call   1460 (2.63%)
    jmp    1423 (2.56%)
    cmp    1349 (2.43%)
    push   1112 (2.00%)
    jne    1043 (1.88%)
    pop    1041 (1.87%)
    test   888 (1.60%)
    movsxd 432 (0.78%)
    sub    431 (0.78%)
    ret    414 (0.74%)
    inc    394 (0.71%)
    imul   326 (0.59%)
    shl    267 (0.48%)
    xor    253 (0.46%)
    jle    183 (0.33%)
    jbe    182 (0.33%)
    jae    167 (0.30%)
    ja     166 (0.30%)
    jb     163 (0.29%)
    and    150 (0.27%)
    shr    132 (0.24%)
    dec    129 (0.23%)
    jge    103 (0.19%)
    jg     93 (0.17%)
    or     92 (0.17%)
    jl     75 (0.13%)
    neg    71 (0.13%)
    sar    61 (0.11%)
    js     30 (0.05%)
    jp     27 (0.05%)
    adc    21 (0.04%)
    cdqe   17 (0.03%)
    jnp    16 (0.03%)
    movsx  10 (0.02%)
    shld   10 (0.02%)
    jns    3 (0.01%)
    bt     2 (0.00%)
    mul    2 (0.00%)
    not    2 (0.00%)
    div    1 (0.00%)
    jno    1 (0.00%)
    jo     1 (0.00%)

INTEL386 10253 (18.44%)
    mov   3550 (6.38%)
    cmp   1225 (2.20%)
    xor   1187 (2.13%)
    movzx 1036 (1.86%)
    sub   449 (0.81%)
    test  449 (0.81%)
    add   396 (0.71%)
    lea   295 (0.53%)
    shr   227 (0.41%)
    and   214 (0.38%)
    sar   194 (0.35%)
    movsx 150 (0.27%)
    sete  136 (0.24%)
    imul  127 (0.23%)
    shl   127 (0.23%)
    inc   88 (0.16%)
    neg   78 (0.14%)
    or    75 (0.13%)
    setne 66 (0.12%)
    dec   29 (0.05%)
    setb  24 (0.04%)
    setg  22 (0.04%)
    not   18 (0.03%)
    setl  16 (0.03%)
    div   14 (0.03%)
    setae 10 (0.02%)
    sbb   7 (0.01%)
    setge 6 (0.01%)
    idiv  6 (0.01%)
    cdq   6 (0.01%)
    seta  5 (0.01%)
    bt    4 (0.01%)
    btc   4 (0.01%)
    setle 4 (0.01%)
    setbe 2 (0.00%)
    setns 2 (0.00%)
    rol   1 (0.00%)
    setnp 1 (0.00%)
    cwde  1 (0.00%)
    sets  1 (0.00%)
    adc   1 (0.00%)

AVX 8653 (15.56%)
    vmovdqa        1156 (2.08%)
    vmovsd         758 (1.36%)
    vmovdqu        457 (0.82%)
    vmulsd         410 (0.74%)
    vpsubw         330 (0.59%)
    vpaddd         309 (0.56%)
    vzeroupper     302 (0.54%)
    vmovapd        290 (0.52%)
    vpaddw         270 (0.49%)
    vpshufd        241 (0.43%)
    vmovd          220 (0.40%)
    vmovups        207 (0.37%)
    vmovupd        191 (0.34%)
    vcvtsi2sd      189 (0.34%)
    vpmaddwd       184 (0.33%)
    vpunpcklwd     166 (0.30%)
    vmovq          163 (0.29%)
    vxorpd         155 (0.28%)
    vucomisd       146 (0.26%)
    vpmaxub        112 (0.20%)
    vdivsd         109 (0.20%)
    vaddsd         108 (0.19%)
    vpslldq        93 (0.17%)
    vpsrlw         93 (0.17%)
    vpxor          93 (0.17%)
    vpunpckhwd     92 (0.17%)
    vpmaxsw        90 (0.16%)
    vpalignr       89 (0.16%)
    vpminsw        84 (0.15%)
    vpor           83 (0.15%)
    vsubsd         82 (0.15%)
    vpsrldq        81 (0.15%)
    vcvttsd2si     72 (0.13%)
    vmovaps        71 (0.13%)
    vmulpd         66 (0.12%)
    vpextrw        61 (0.11%)
    vpmaxsd        59 (0.11%)
    vmovddup       59 (0.11%)
    vpminsd        58 (0.10%)
    vshufpd        53 (0.10%)
    vpinsrd        46 (0.08%)
    vcvtsd2si      41 (0.07%)
    vxorps         38 (0.07%)
    vcvtdq2pd      37 (0.07%)
    vpsrad         33 (0.06%)
    vpunpcklqdq    33 (0.06%)
    vaddpd         31 (0.06%)
    vgf2p8affineqb 30 (0.05%)
    vmaxsd         29 (0.05%)
    vpand          27 (0.05%)
    vpextrd        26 (0.05%)
    vpsrld         25 (0.04%)
    vpmulhw        24 (0.04%)
    vpshufb        24 (0.04%)
    vandpd         23 (0.04%)
    vpsubd         23 (0.04%)
    vpmullw        22 (0.04%)
    vpackuswb      20 (0.04%)
    vpsraw         18 (0.03%)
    vsqrtsd        18 (0.03%)
    vpackssdw      17 (0.03%)
    vcvttpd2dq     16 (0.03%)
    vpunpcklbw     16 (0.03%)
    vpinsrw        16 (0.03%)
    vpunpckhbw     15 (0.03%)
    vdivpd         10 (0.02%)
    vsubpd         9 (0.02%)
    vpsllw         9 (0.02%)
    vpcmpeqd       8 (0.01%)
    vshufps        8 (0.01%)
    vpmovsxbw      8 (0.01%)
    vpmovzxbw      7 (0.01%)
    vminsd         7 (0.01%)
    vblendpd       7 (0.01%)
    vextractf128   6 (0.01%)
    vinsertf128    6 (0.01%)
    vmovlpd        6 (0.01%)
    vpabsw         6 (0.01%)
    vpmovsxbd      6 (0.01%)
    vbroadcastsd   6 (0.01%)
    vperm2f128     5 (0.01%)
    vpabsd         4 (0.01%)
    vinsertps      4 (0.01%)
    vmaxpd         4 (0.01%)
    vmovhpd        3 (0.01%)
    vpunpckldq     3 (0.01%)
    vpavgw         3 (0.01%)
    vpblendw       3 (0.01%)
    vmovlps        2 (0.00%)
    vpmovsxbq      2 (0.00%)
    vpaddusb       2 (0.00%)
    vextractps     1 (0.00%)
    vpslld         1 (0.00%)
    vpcmpeqb       1 (0.00%)
    vpsllq         1 (0.00%)
    vpaddb         1 (0.00%)
    vpermilpd      1 (0.00%)
    vpextrb        1 (0.00%)
    vbroadcastf128 1 (0.00%)
    vpavgb         1 (0.00%)

INTEL8086 4095 (7.36%)
    int3 2328 (4.19%)
    cmp  491 (0.88%)
    mov  463 (0.83%)
    test 350 (0.63%)
    add  147 (0.26%)
    nop  99 (0.18%)
    and  71 (0.13%)
    or   56 (0.10%)
    sub  51 (0.09%)
    shr  23 (0.04%)
    xor  9 (0.02%)
    dec  3 (0.01%)
    inc  2 (0.00%)
    adc  1 (0.00%)
    not  1 (0.00%)

AVX2 2575 (4.63%)
    vpaddd         299 (0.54%)
    vpsubw         209 (0.38%)
    vpaddw         194 (0.35%)
    vpmovsxwd      124 (0.22%)
    vpmaddwd       123 (0.22%)
    vpsrad         88 (0.16%)
    vpalignr       87 (0.16%)
    vpminsw        80 (0.14%)
    vpsrlw         78 (0.14%)
    vpminsd        69 (0.12%)
    vpmaxsd        66 (0.12%)
    vextracti128   64 (0.12%)
    vpmaxub        64 (0.12%)
    vpmulld        62 (0.11%)
    vpunpcklwd     62 (0.11%)
    vpunpckhwd     61 (0.11%)
    vpmaxsw        60 (0.11%)
    vpbroadcastd   59 (0.11%)
    vpbroadcastw   55 (0.10%)
    vpermq         53 (0.10%)
    vpslld         52 (0.09%)
    vpsrld         47 (0.08%)
    vpsubd         35 (0.06%)
    vpshufd        32 (0.06%)
    vpbroadcastq   32 (0.06%)
    vpaddq         30 (0.05%)
    vpsllq         26 (0.05%)
    vinserti128    25 (0.04%)
    vpabsw         23 (0.04%)
    vperm2i128     23 (0.04%)
    vpmovzxbw      21 (0.04%)
    vbroadcastsd   20 (0.04%)
    vpxor          19 (0.03%)
    vpackssdw      17 (0.03%)
    vpblendd       16 (0.03%)
    vpackusdw      15 (0.03%)
    vpmullw        15 (0.03%)
    vpmovzxbd      14 (0.03%)
    vpand          13 (0.02%)
    vpackuswb      12 (0.02%)
    vpunpcklbw     12 (0.02%)
    vpunpckhbw     11 (0.02%)
    vpermd         11 (0.02%)
    vpermpd        10 (0.02%)
    vpor           9 (0.02%)
    vpaddb         8 (0.01%)
    vpcmpeqd       8 (0.01%)
    vpblendw       6 (0.01%)
    vpmovzxwd      6 (0.01%)
    vpmulhw        6 (0.01%)
    vpsraw         6 (0.01%)
    vpsrldq        6 (0.01%)
    vbroadcasti128 5 (0.01%)
    vpsllw         5 (0.01%)
    vpavgb         4 (0.01%)
    vpsravd        4 (0.01%)
    vpsllvd        3 (0.01%)
    vpavgw         3 (0.01%)
    vpsrlvd        2 (0.00%)
    vpaddusb       2 (0.00%)
    vpunpcklqdq    2 (0.00%)
    vpcmpeqb       1 (0.00%)
    vpslldq        1 (0.00%)

AVX512VL 1035 (1.86%)
    vpbroadcastd  100 (0.18%)
    vpaddd        83 (0.15%)
    vpbroadcastw  51 (0.09%)
    vmovdqa64     39 (0.07%)
    vmovdqu64     37 (0.07%)
    vpgatherqd    37 (0.07%)
    vpsubw        34 (0.06%)
    vpmovdw       32 (0.06%)
    vpermt2d      31 (0.06%)
    vpmaxsd       30 (0.05%)
    vmovdqu16     28 (0.05%)
    vpmovwb       28 (0.05%)
    vpminsd       27 (0.05%)
    vpscatterqd   26 (0.05%)
    vpcmpub       25 (0.04%)
    vandpd        23 (0.04%)
    vpmovsxwd     20 (0.04%)
    vpshufd       20 (0.04%)
    vpsubb        20 (0.04%)
    vpmovuswb     18 (0.03%)
    vextracti32x4 17 (0.03%)
    vpshufb       17 (0.03%)
    vmovapd       16 (0.03%)
    vcmppd        15 (0.03%)
    vpermt2w      14 (0.03%)
    vpmovdb       14 (0.03%)
    vpbroadcastq  14 (0.03%)
    vpermt2b      13 (0.02%)
    vcvtpd2qq     10 (0.02%)
    vmovupd       10 (0.02%)
    vxorpd        10 (0.02%)
    vpmovqd       9 (0.02%)
    vfmadd213pd   8 (0.01%)
    vinserti32x4  8 (0.01%)
    vpslld        8 (0.01%)
    vpxord        8 (0.01%)
    vpandd        6 (0.01%)
    vpscatterdd   6 (0.01%)
    vpcmpuq       6 (0.01%)
    vmulpd        6 (0.01%)
    vpermw        5 (0.01%)
    vpgatherdd    5 (0.01%)
    vpmovm2b      5 (0.01%)
    vpmovusdb     5 (0.01%)
    vpcmpgtd      5 (0.01%)
    vpmovzxbd     5 (0.01%)
    vpord         5 (0.01%)
    vpermi2q      5 (0.01%)
    vaddpd        4 (0.01%)
    vfmadd231pd   4 (0.01%)
    vpermi2d      4 (0.01%)
    vpmulld       4 (0.01%)
    vpternlogq    4 (0.01%)
    vpmovd2m      3 (0.01%)
    vpermb        3 (0.01%)
    vpermt2pd     3 (0.01%)
    vpsrad        3 (0.01%)
    vpternlogd    3 (0.01%)
    valignd       2 (0.00%)
    vpcmpud       2 (0.00%)
    vpalignr      2 (0.00%)
    vpmovm2d      2 (0.00%)
    vpermi2pd     2 (0.00%)
    vmovdqu32     2 (0.00%)
    vpbroadcastb  2 (0.00%)
    vcvtudq2pd    2 (0.00%)
    vcvtqq2pd     2 (0.00%)
    vpsubd        2 (0.00%)
    vpmaxsq       2 (0.00%)
    vshufpd       2 (0.00%)
    vpminsq       2 (0.00%)
    vpermi2w      1 (0.00%)
    vpandnd       1 (0.00%)
    vpblendmb     1 (0.00%)
    valignq       1 (0.00%)
    vpcmpd        1 (0.00%)
    vptestmd      1 (0.00%)
    vptestnmd     1 (0.00%)
    vpunpcklqdq   1 (0.00%)
    vorpd         1 (0.00%)
    vsubpd        1 (0.00%)

AVX512F 926 (1.67%)
    vpbroadcastd  100 (0.18%)
    vpaddd        83 (0.15%)
    kxnorw        63 (0.11%)
    vmovdqa64     39 (0.07%)
    vmovdqu64     37 (0.07%)
    vpgatherqd    37 (0.07%)
    vpmovdw       32 (0.06%)
    vmovsd        31 (0.06%)
    vpermt2d      31 (0.06%)
    vpmaxsd       30 (0.05%)
    vpminsd       27 (0.05%)
    vpscatterqd   26 (0.05%)
    vaddsd        21 (0.04%)
    vpshufd       20 (0.04%)
    vpmovsxwd     20 (0.04%)
    vmulsd        18 (0.03%)
    vextracti32x4 17 (0.03%)
    vmovapd       16 (0.03%)
    vcmppd        15 (0.03%)
    vpbroadcastq  14 (0.03%)
    vpmovdb       14 (0.03%)
    vmovupd       10 (0.02%)
    vucomisd      10 (0.02%)
    vfmadd213sd   9 (0.02%)
    vpmovqd       9 (0.02%)
    vscalefsd     8 (0.01%)
    vfmadd213pd   8 (0.01%)
    vinserti32x4  8 (0.01%)
    vpslld        8 (0.01%)
    vpxord        8 (0.01%)
    vfmadd231sd   7 (0.01%)
    vcmpsd        7 (0.01%)
    vpandd        6 (0.01%)
    korw          6 (0.01%)
    vmulpd        6 (0.01%)
    vpscatterdd   6 (0.01%)
    vcvtsi2sd     6 (0.01%)
    vpcmpuq       6 (0.01%)
    vcvtusi2sd    5 (0.01%)
    vpermi2q      5 (0.01%)
    vpgatherdd    5 (0.01%)
    knotw         5 (0.01%)
    vpmovusdb     5 (0.01%)
    vpmovzxbd     5 (0.01%)
    vpord         5 (0.01%)
    vpcmpgtd      5 (0.01%)
    vfmadd231pd   4 (0.01%)
    vaddpd        4 (0.01%)
    vpmulld       4 (0.01%)
    vpternlogq    4 (0.01%)
    vpermi2d      4 (0.01%)
    vpermt2pd     3 (0.01%)
    vfmsub231sd   3 (0.01%)
    vfnmadd213sd  3 (0.01%)
    vpsrad        3 (0.01%)
    vpternlogd    3 (0.01%)
    vpermi2pd     2 (0.00%)
    vmovdqu32     2 (0.00%)
    vcvtudq2pd    2 (0.00%)
    vpmaxsq       2 (0.00%)
    vpminsq       2 (0.00%)
    vpsubd        2 (0.00%)
    vpcmpud       2 (0.00%)
    valignd       2 (0.00%)
    kandw         2 (0.00%)
    vshufpd       2 (0.00%)
    vfmsub213sd   2 (0.00%)
    kmovw         1 (0.00%)
    vpandnd       1 (0.00%)
    valignq       1 (0.00%)
    vpcmpd        1 (0.00%)
    vfnmadd231sd  1 (0.00%)
    vdivsd        1 (0.00%)
    vptestmd      1 (0.00%)
    vptestnmd     1 (0.00%)
    vpunpcklqdq   1 (0.00%)
    vsubpd        1 (0.00%)

CMOV 790 (1.42%)
    cmovae 281 (0.51%)
    cmovge 84 (0.15%)
    cmove  74 (0.13%)
    cmovl  74 (0.13%)
    cmovg  65 (0.12%)
    cmovne 56 (0.10%)
    cmovs  49 (0.09%)
    cmovb  26 (0.05%)
    cmovbe 26 (0.05%)
    cmova  19 (0.03%)
    cmovle 18 (0.03%)
    cmovp  12 (0.02%)
    cmovns 5 (0.01%)
    cmovo  1 (0.00%)

MULTIBYTENOP 597 (1.07%)
    nop 597 (1.07%)

AVX512BW 282 (0.51%)
    vpbroadcastw 51 (0.09%)
    vpsubw       34 (0.06%)
    vmovdqu16    28 (0.05%)
    vpmovwb      28 (0.05%)
    vpcmpub      25 (0.04%)
    kmovd        24 (0.04%)
    vpsubb       20 (0.04%)
    vpmovuswb    18 (0.03%)
    vpshufb      17 (0.03%)
    vpermt2w     14 (0.03%)
    kmovq        7 (0.01%)
    vpermw       5 (0.01%)
    vpmovm2b     5 (0.01%)
    vpalignr     2 (0.00%)
    vpbroadcastb 2 (0.00%)
    vpblendmb    1 (0.00%)
    vpermi2w     1 (0.00%)

FMA 226 (0.41%)
    vfmadd231sd  86 (0.15%)
    vfmadd213sd  54 (0.10%)
    vfmadd132sd  19 (0.03%)
    vfmadd213pd  17 (0.03%)
    vfnmadd213sd 14 (0.03%)
    vfnmadd231sd 11 (0.02%)
    vfmadd231pd  7 (0.01%)
    vfmsub231sd  6 (0.01%)
    vfmadd132pd  5 (0.01%)
    vfnmadd213pd 5 (0.01%)
    vfmsub213sd  2 (0.00%)

BMI2 154 (0.28%)
    mulx 102 (0.18%)
    shlx 44 (0.08%)
    sarx 4 (0.01%)
    shrx 4 (0.01%)

AVX512DQ 62 (0.11%)
    vandpd    23 (0.04%)
    vcvtpd2qq 10 (0.02%)
    vxorpd    10 (0.02%)
    korb      3 (0.01%)
    kshiftlb  3 (0.01%)
    kxnorb    3 (0.01%)
    vpmovd2m  3 (0.01%)
    kshiftrb  2 (0.00%)
    vcvtqq2pd 2 (0.00%)
    vpmovm2d  2 (0.00%)
    vorpd     1 (0.00%)

GFNI 30 (0.05%)
    vgf2p8affineqb 30 (0.05%)

BMI1 23 (0.04%)
    andn 23 (0.04%)

AVX512_VBMI 16 (0.03%)
    vpermt2b 13 (0.02%)
    vpermb   3 (0.01%)

INTEL486 9 (0.02%)
    bswap 9 (0.02%)

MOVBE 9 (0.02%)
    movbe 9 (0.02%)

LZCNT 3 (0.01%)
    lzcnt 3 (0.01%)

CET_IBT 2 (0.00%)
    endbr64 2 (0.00%)

CPUID 1 (0.00%)
    cpuid 1 (0.00%)

XSAVE 1 (0.00%)
    xgetbv 1 (0.00%)

Registers
---------
= 69484

RAX   4622 (6.65%)
XMM0  4192 (6.03%)
EAX   3206 (4.61%)
RDI   2741 (3.94%)
RCX   2596 (3.74%)
XMM1  2415 (3.48%)
RDX   2017 (2.90%)
XMM2  2007 (2.89%)
XMM3  1905 (2.74%)
RSI   1883 (2.71%)
RBX   1878 (2.70%)
ECX   1747 (2.51%)
EDX   1706 (2.46%)
R14   1635 (2.35%)
R15   1586 (2.28%)
XMM4  1487 (2.14%)
YMM0  1409 (2.03%)
ESI   1343 (1.93%)
R12   1338 (1.93%)
R9    1215 (1.75%)
R8    1121 (1.61%)
R13   1088 (1.57%)
XMM5  1058 (1.52%)
YMM1  1003 (1.44%)
EBP   984 (1.42%)
YMM4  912 (1.31%)
R9D   866 (1.25%)
XMM6  840 (1.21%)
RBP   838 (1.21%)
YMM3  826 (1.19%)
YMM2  823 (1.18%)
R8D   797 (1.15%)
R14D  720 (1.04%)
YMM5  680 (0.98%)
EDI   677 (0.97%)
R11   609 (0.88%)
XMM7  579 (0.83%)
R10   562 (0.81%)
EBX   559 (0.80%)
AL    546 (0.79%)
YMM6  545 (0.78%)
R10D  518 (0.75%)
R15D  506 (0.73%)
R12D  488 (0.70%)
R11D  464 (0.67%)
XMM8  464 (0.67%)
R13D  420 (0.60%)
YMM7  389 (0.56%)
YMM8  382 (0.55%)
DL    370 (0.53%)
XMM9  359 (0.52%)
RSP   346 (0.50%)
XMM10 314 (0.45%)
YMM9  306 (0.44%)
XMM11 271 (0.39%)
CL    255 (0.37%)
YMM10 243 (0.35%)
XMM15 241 (0.35%)
YMM11 235 (0.34%)
SIL   218 (0.31%)
YMM12 217 (0.31%)
K0    191 (0.27%)
YMM13 187 (0.27%)
YMM14 168 (0.24%)
XMM12 155 (0.22%)
XMM19 144 (0.21%)
K1    137 (0.20%)
BPL   102 (0.15%)
XMM13 101 (0.15%)
R10L  97 (0.14%)
YMM19 94 (0.14%)
R9L   87 (0.13%)
YMM15 82 (0.12%)
CX    78 (0.11%)
R8L   74 (0.11%)
XMM20 73 (0.11%)
XMM14 71 (0.10%)
DIL   67 (0.10%)
BL    54 (0.08%)
R11L  51 (0.07%)
YMM16 48 (0.07%)
SI    44 (0.06%)
XMM18 43 (0.06%)
AX    41 (0.06%)
YMM21 39 (0.06%)
YMM17 37 (0.05%)
R8W   36 (0.05%)
XMM16 35 (0.05%)
R13L  34 (0.05%)
R15L  32 (0.05%)
XMM17 32 (0.05%)
YMM20 30 (0.04%)
BP    29 (0.04%)
YMM18 29 (0.04%)
R12L  28 (0.04%)
R14W  28 (0.04%)
DX    27 (0.04%)
YMM23 26 (0.04%)
K2    25 (0.04%)
YMM22 23 (0.03%)
YMM25 22 (0.03%)
XMM24 21 (0.03%)
YMM24 17 (0.02%)
R14L  17 (0.02%)
R9W   17 (0.02%)
R10W  16 (0.02%)
YMM26 15 (0.02%)
XMM23 14 (0.02%)
XMM21 13 (0.02%)
XMM22 11 (0.02%)
BX    10 (0.01%)
XMM25 9 (0.01%)
R11W  9 (0.01%)
R12W  9 (0.01%)
XMM31 6 (0.01%)
K3    6 (0.01%)
DI    5 (0.01%)
R13W  5 (0.01%)
K4    4 (0.01%)
XMM26 3 (0.00%)
R15W  3 (0.00%)
AH    2 (0.00%)
XMM27 2 (0.00%)
DH    1 (0.00%)
BH    1 (0.00%)



</p>
</details>

rcombs commented:
> this fixes an edge issue in a special case where libass is built with -mavx but not -mavx2, and vzeroupper has been disabled.

Sorry, what's the issue in question? Can this crash or something?

> this also reduces the number of sse instructions in the final build product targeted for avx2+.

I'm a little confused as to why this is desirable; these code paths won't be taken outside of test situations, right? Why does this matter?

Overall this change seems harmless, I just don't get what it's for.

astiob commented:
I assume this ideally ought to be implemented in the autotools build, too.

Andarwinux commented:
> Sorry, what's the issue in question? Can this crash or something?

avx-sse transition penalty. In this scenario, the target architecture may lack avx2, so it still uses ssse3 and sse2 paths. However, enable avx is sufficient for compilers to generate 256-bit instructions.

> I'm a little confused as to why this is desirable; these code paths won't be taken outside of test situations, right? Why does this matter?

Honestly just OCD. I would use bin-cpuflags-x86 to verify what instructions compilers issued for binaries, and in this case some sse instructions that will never be executed would interfere with the stats and be a bit “ugly” because ideally avx2-targeted binaries shouldn't include any sse instructions.

> I assume this ideally ought to be implemented in the autotools build, too.

It looks like autotools doesn't have an equivalent of get_define, and it looks too ugly to emulate it with AC_COMPILE_IFELSE.

rcombs commented:
Is there actually a measurable penalty from using non-VEX instructions on any real CPUs that support AVX but not AVX2?

Andarwinux commented:
> Is there actually a measurable penalty from using non-VEX instructions on any real CPUs that support AVX but not AVX2?

This is true if vzeroupper is disabled.

TheOneric commented:
> It looks like autotools doesn't have an equivalent of get_define, and it looks too ugly to emulate it with `AC_COMPILE_IFELSE`.

If you only care about the macro being defined, but not its value I believe `AC_CHECK_DECL` will work and is concise.
If you care about the value to `AC_COMPILE_IFELSE` will work and isn’ŧ that ugly imho, but better yet you can use `AC_PREPROC_IFELSE` to avoid wasting time on running the compile step.

Generally, regardless of whether you personally believe it ugly or not, any such build change MUST be properly implemented in autotools; leaving the implementation incomplete like this is unacceptable and definitely a blocker.

> > Is there actually a measurable penalty from using non-VEX instructions on any real CPUs that support AVX but not AVX2?
>
> This is true if vzeroupper is disabled.

Could you provide some benchmarks for reference? So we can better understand the impact

Andarwinux commented:
> Could you provide some benchmarks for reference? So we can better understand the impact



| -O3 -mavx -mtune=skylake -mno-vzeroupper | -O3 -mavx -mtune=skylake | -O3 -mavx -mtune=skylake -mno-vzeroupper SSE2AVX |
| -- | -- | -- |
|nop: 20.0  |     nop: 20.0 |      nop: 20.0|
|add_bitmaps_c: 116.9 |   add_bitmaps_c: 117.0 |   add_bitmaps_c: 113.0|
add_bitmaps_sse2: 34.7|  add_bitmaps_sse2: 29.5 | add_bitmaps_sse2: 28.7|
add_bitmaps_avx2: 21.5  |add_bitmaps_avx2: 15.9 | add_bitmaps_avx2: 15.1|
be_blur_c: 985.8        |be_blur_c: 987.9 |       be_blur_c: 1155.8|
be_blur_sse2: 152.7   |  be_blur_sse2: 157.5|     be_blur_sse2: 149.7|
be_blur_ssse3: 117.3   | be_blur_ssse3: 119.8|    be_blur_ssse3: 108.9|
be_blur_avx2: 95.3      |be_blur_avx2: 94.8   |   be_blur_avx2: 95.0|
blur4_horz16_c: 4767.4|  blur4_horz16_c: 4760.5 | blur4_horz16_c: 4873.5|
blur4_horz16_sse2: 1267.4 |      blur4_horz16_sse2: 1274.0     |  blur4_horz16_sse2: 1258.2|
blur4_horz16_ssse3: 1109.8 |      blur4_horz16_ssse3: 1119.9    |  blur4_horz16_ssse3: 1122.7|
blur4_horz32_c: 3988.4 | blur4_horz32_c: 3990.6  | blur4_horz32_c: 4002.1 |
blur4_horz32_avx2: 655.1|        blur4_horz32_avx2: 656.7  |      blur4_horz32_avx2: 654.2|
blur4_vert16_c: 3919.1 | blur4_vert16_c: 3932.6  | blur4_vert16_c: 3881.1|
blur4_vert16_sse2: 1741.2 |      blur4_vert16_sse2: 1752.6  |     blur4_vert16_sse2: 1738.2|
blur4_vert32_c: 3790.3  |blur4_vert32_c: 3792.7 | blur4_vert32_c: 3775.5|
blur4_vert32_avx2: 888.7|        blur4_vert32_avx2: 882.0  |      blur4_vert32_avx2: 881.4|
blur5_horz16_c: 5713.4  | blur5_horz16_c: 5713.8  | blur5_horz16_c: 5679.2|
blur5_horz16_sse2: 1965.3 |       blur5_horz16_sse2: 1980.2  |     blur5_horz16_sse2: 1939.9|
blur5_horz16_ssse3: 1623.1 |      blur5_horz16_ssse3: 1636.1   |   blur5_horz16_ssse3: 1591.9|
blur5_horz32_c: 4613.7 | blur5_horz32_c: 4619.5 | blur5_horz32_c: 4627.4|
blur5_horz32_avx2: 796.0       | blur5_horz32_avx2: 786.6   |     blur5_horz32_avx2: 784.1|
blur5_vert16_c: 5084.5 |  blur5_vert16_c: 5091.0  | blur5_vert16_c: 5054.6|
blur5_vert16_sse2: 2291.0       | blur5_vert16_sse2: 2307.5  |     blur5_vert16_sse2: 2295.1|
blur5_vert32_c: 4832.2 | blur5_vert32_c: 4833.0 | blur5_vert32_c: 4804.0|
blur5_vert32_avx2: 1175.2       | blur5_vert32_avx2: 1167.7  |     blur5_vert32_avx2: 1160.6|
blur6_horz16_c: 6163.1  | blur6_horz16_c: 6166.0 | blur6_horz16_c: 6085.0|
blur6_horz16_sse2: 2277.5  |     blur6_horz16_sse2: 2292.6 |      blur6_horz16_sse2: 2251.5 |
blur6_horz16_ssse3: 1901.2 |     blur6_horz16_ssse3: 1906.9 |     blur6_horz16_ssse3: 1872.7 |
blur6_horz32_c: 5132.1 | blur6_horz32_c: 5130.8  | blur6_horz32_c: 5145.6 |
blur6_horz32_avx2: 926.9  |      blur6_horz32_avx2: 915.2  |      blur6_horz32_avx2: 914.2 |
blur6_vert16_c: 6543.9  | blur6_vert16_c: 6557.8  | blur6_vert16_c: 6479.5 |
blur6_vert16_sse2: 2904.5 |      blur6_vert16_sse2: 2921.3  |     blur6_vert16_sse2: 2894.6 |
blur6_vert32_c: 5965.1  | blur6_vert32_c: 5967.3  | blur6_vert32_c: 5939.6 |
blur6_vert32_avx2: 1488.3 |      blur6_vert32_avx2: 1469.5   |    blur6_vert32_avx2: 1468.3 |
blur7_horz16_c: 6671.7  | blur7_horz16_c: 6763.5  | blur7_horz16_c: 6666.8 |
blur7_horz16_sse2: 2645.1 |      blur7_horz16_sse2: 2659.1    |   blur7_horz16_sse2: 2615.1 |
blur7_horz16_ssse3: 2166.1 |     blur7_horz16_ssse3: 2184.8  |    blur7_horz16_ssse3: 2158.8 |
blur7_horz32_c: 5650.2  | blur7_horz32_c: 5665.4 |  blur7_horz32_c: 5665.6 |
blur7_horz32_avx2: 1055.5  |     blur7_horz32_avx2: 1041.1     |  blur7_horz32_avx2: 1039.1 |
blur7_vert16_c: 8095.1  | blur7_vert16_c: 8120.1 |  blur7_vert16_c: 8039.5 |
blur7_vert16_sse2: 3599.8    |   blur7_vert16_sse2: 3618.9  |     blur7_vert16_sse2: 3600.7 |
blur7_vert32_c: 7232.1 | blur7_vert32_c: 7256.6 | blur7_vert32_c: 7205.3 |
blur7_vert32_avx2: 1842.0    |   blur7_vert32_avx2: 1829.3     |  blur7_vert32_avx2: 1823.8 |
blur8_horz16_c: 6787.2 |  blur8_horz16_c: 6789.3  |blur8_horz16_c: 6777.9|
blur8_horz16_sse2: 2793.7  |     blur8_horz16_sse2: 2805.5    |   blur8_horz16_sse2: 2767.1 |
blur8_horz16_ssse3: 2183.5 |     blur8_horz16_ssse3: 2185.5  |    blur8_horz16_ssse3: 2177.6 |
blur8_horz32_c: 6267.4  | blur8_horz32_c: 6301.4  | blur8_horz32_c: 6281.8 |
blur8_horz32_avx2: 1111.4 |      blur8_horz32_avx2: 1100.5    |   blur8_horz32_avx2: 1094.4 |
blur8_vert16_c: 9772.5  |blur8_vert16_c: 9776.4 | blur8_vert16_c: 9686.3 |
blur8_vert16_sse2: 4361.6  |     blur8_vert16_sse2: 4376.3   |    blur8_vert16_sse2: 4363.0 |
blur8_vert32_c: 8648.0  |blur8_vert32_c: 8709.3 |  blur8_vert32_c: 8607.8 |
blur8_vert32_avx2: 2223.8 |      blur8_vert32_avx2: 2207.3 |      blur8_vert32_avx2: 2203.4 |
expand_horz16_c: 688.5 | expand_horz16_c: 687.2 |  expand_horz16_c: 724.6 |
expand_horz16_sse2: 551.7  |     expand_horz16_sse2: 557.0    |   expand_horz16_sse2: 551.5 |
expand_horz16_ssse3: 531.7  |    expand_horz16_ssse3: 529.6  |    expand_horz16_ssse3: 523. 6 |
expand_horz32_c: 536.3  | expand_horz32_c: 537.3 | expand_horz32_c: 537.5 |
expand_horz32_avx2: 319.3  |     expand_horz32_avx2: 321.0   |    expand_horz32_avx2: 319.8 |
expand_vert16_c: 3563.7  |expand_vert16_c: 3563.9 | expand_vert16_c: 3524.5 |
expand_vert16_sse2: 379.5  |     expand_vert16_sse2: 382.1   |    expand_vert16_sse2: 383.0 |
expand_vert32_c: 529.3  |expand_vert32_c: 530.4 | expand_vert32_c: 528.5 |
expand_vert32_avx2: 191.2 |      expand_vert32_avx2: 193.8   |    expand_vert32_avx2: 193.0 |
fill_generic_tile16_c: 271.1   | fill_generic_tile16_c: 267.7  |  fill_generic_tile16_c: 267.0 |
fill_generic_tile16_sse2: 129.3 | fill_generic_tile16_sse2: 130.5 | fill_generic_tile16_sse2: 139.1 |
fill_generic_tile16_ssse3: 124.8   |     fill_generic_tile16_ssse3: 125.9  |      fill_generic_tile16_ssse3: 131.4 |
fill_generic_tile16_avx2: 99.2 | fill_generic_tile16_avx2: 100.3 | fill_generic_tile16_avx2: 109.4 |
fill_generic_tile32_c: 480.5   | fill_generic_tile32_c: 483.5   | fill_generic_tile32_c: 490.1 |
fill_generic_tile32_sse2: 271.8 | fill_generic_tile32_sse2: 277.7 | fill_generic_tile32_sse2: 276.4 |
fill_generic_tile32_ssse3: 247.7 |        fill_generic_tile32_ssse3: 249.0    |    fill_generic_tile32_ssse3: 250.5 |
fill_generic_tile32_avx2: 171.7 | fill_generic_tile32_avx2: 171.2 | fill_generic_tile32_avx2: 179.0 |
fill_halfplane_tile16_c: 89.6  | fill_halfplane_tile16_c: 87.9  | fill_halfplane_tile16_c: 91.5 |
fill_halfplane_tile16_sse2: 68.9  |      fill_halfplane_tile16_sse2: 67.6  |      fill_halfplane_tile16_sse2: 67.2 |
fill_halfplane_tile16_avx2: 44.1    |    fill_halfplane_tile16_avx2: 44.5   |     fill_halfplane_tile16_avx2: 45.4 |
fill_halfplane_tile32_c: 303.3  | fill_halfplane_tile32_c: 303.6 | fill_halfplane_tile32_c: 324.8 |
fill_halfplane_tile32_sse2: 249.0  |     fill_halfplane_tile32_sse2: 248.2  |     fill_halfplane_tile32_sse2: 239.4 |
fill_halfplane_tile32_avx2: 137.8    |   fill_halfplane_tile32_avx2: 135.5     |  fill_halfplane_tile32_avx2: 136.0 |
fill_solid_tile16_c: 5.3     |   fill_solid_tile16_c: 5.1   |    | fill_solid_tile16_c: 5.2 |
fill_solid_tile16_sse2: 4.9  |   fill_solid_tile16_sse2: 5.1   |  fill_solid_tile16_sse2: 4.9 |
fill_solid_tile16_avx2: 4.8   |  fill_solid_tile16_avx2: 4.7  |    fill_solid_tile16_avx2: 4.8 |
fill_solid_tile32_c: 11.1     |  fill_solid_tile32_c: 10.8     |  fill_solid_tile32_c: 10.9 |
fill_solid_tile32_sse2: 15.7  |  fill_solid_tile32_sse2: 16.1   | fill_solid_tile32_sse2: 15.7 |
fill_solid_tile32_avx2: 10.9 |   fill_solid_tile32_avx2: 11.4 |   fill_solid_tile32_avx2: 11.1 |
imul_bitmaps_c: 75.3   | imul_bitmaps_c: 75.2   | imul_bitmaps_c: 74.7 | 
imul_bitmaps_sse2: 73.3 | imul_bitmaps_sse2: 79.6 | imul_bitmaps_sse2: 79.5 |
imul_bitmaps_avx2: 40.2 | imul_bitmaps_avx2: 39.5 | imul_bitmaps_avx2: 38.8 |
merge_tile16_c: 9.2    | merge_tile16_c: 9.3   |  merge_tile16_c: 9.5 |
merge_tile16_sse2: 10.0| merge_tile16_sse2: 10.1 | merge_tile16_sse2: 9.8 |
merge_tile16_avx2: 9.8 | merge_tile16_avx2: 10.0 | merge_tile16_avx2: 10.0 |
merge_tile32_c: 43.2   | merge_tile32_c: 43.2  |  merge_tile32_c: 43.1 |
merge_tile32_sse2: 43.8 | merge_tile32_sse2: 44.3 | merge_tile32_sse2: 44.0 |
merge_tile32_avx2: 18.9 | merge_tile32_avx2: 19.3 | merge_tile32_avx2: 19.2 |
mul_bitmaps_c: 72.7    | mul_bitmaps_c: 73.4    | mul_bitmaps_c: 72.8 |
mul_bitmaps_sse2: 71.6 | mul_bitmaps_sse2: 72.6 | mul_bitmaps_sse2: 70.3 |
mul_bitmaps_avx2: 39.4 | mul_bitmaps_avx2: 39.4 | mul_bitmaps_avx2: 39.3 |
shrink_horz16_c: 1109.8 | shrink_horz16_c: 1111.5 | shrink_horz16_c: 1149.2 |
shrink_horz16_sse2: 581.0      | shrink_horz16_sse2: 586.9     |  shrink_horz16_sse2: 579.6 |
shrink_horz16_ssse3: 508.3  |    shrink_horz16_ssse3: 512.6 |     shrink_horz16_ssse3: 503.4 |
shrink_horz32_c: 1317.9 | shrink_horz32_c: 1315.9 | shrink_horz32_c: 1352.8|
shrink_horz32_avx2: 352.8   |    shrink_horz32_avx2: 353.4    |   shrink_horz32_avx2: 354.6|
shrink_vert16_c: 1540.0 | shrink_vert16_c: 1542.4 | shrink_vert16_c: 1538.6|
shrink_vert16_sse2: 335.4  |     shrink_vert16_sse2: 344.0     |  shrink_vert16_sse2: 339.4|
shrink_vert32_c: 612.3  | shrink_vert32_c: 611.4 | shrink_vert32_c: 606.7|
shrink_vert32_avx2: 170.3 |      shrink_vert32_avx2: 170.5     |  shrink_vert32_avx2: 170.3|
stripe_pack16_c: 652.3  | stripe_pack16_c: 652.9  | stripe_pack16_c: 651.2|
stripe_pack16_sse2: 125.1 |      stripe_pack16_sse2: 100.8   |    stripe_pack16_sse2: 123.4
stripe_pack32_c: 626.6 | stripe_pack32_c: 625.9 | stripe_pack32_c: 626.0|
stripe_pack32_avx2: 57.0      |  stripe_pack32_avx2: 55.7       | stripe_pack32_avx2: 55.7|
stripe_unpack16_c: 115.0       | stripe_unpack16_c: 114.0        |stripe_unpack16_c: 114.3|
stripe_unpack16_sse2: 118.4 |    stripe_unpack16_sse2: 119.5 |    stripe_unpack16_sse2: 116.4|
stripe_unpack32_c: 108.6     |   stripe_unpack32_c: 104.4      |  stripe_unpack32_c: 106.0|
stripe_unpack32_avx2: 97.1   |   stripe_unpack32_avx2: 91.4  |    stripe_unpack32_avx2: 97.1|

> Generally, regardless of whether you personally believe it ugly or not, any such build change MUST be properly implemented in autotools; leaving the implementation incomplete like this is unacceptable and definitely a blocker.

I thought autotools was close to obsolete for libass. In my experience as a distro builder, this is true for most projects, once cmake/meson is available, autotools is quickly abandoned.

Andarwinux committed:
build/nasm: auto-convert SSE to AVX when baseline includes AVX

This is handled by an x86inc.asm feature and avoids
penalties from switching between the two formats.

This is only relevant when AVX, but not AVX2 is included in the compiler
baseline as otherwise no SSE2 routines will be selected.

Inspired-by: https://code.videolan.org/videolan/dav1d/-/commit/64c65de3faf1fa5e0272858203a9590f4ca31927

TheOneric commented:
Brinign the checkasm benchmarks into an easier to parse form and filtering it down just to SSE2 routines:

|    | -O3 -mavx -mtune=skylake -mno-vzeroupper | -O3 -mavx -mtune=skylake | -O3 -mavx -mtune=skylake -mno-vzeroupper SSE2AVX |
|---|---|---|---|
|merge_tile16_sse2|10.0|10.1 |9.8 |
|fill_solid_tile16_sse2|4.9  |5.1   |4.9 |
|add_bitmaps_sse2|34.7|29.5 |28.7|
|blur5_vert16_sse2|2291.0       |2307.5  |2295.1|
|stripe_unpack16_sse2|118.4 |119.5 |116.4|
|imul_bitmaps_sse2|73.3 |79.6 |79.5 |
|fill_generic_tile16_sse2|129.3 |130.5 |139.1 |
|blur5_horz16_sse2|1965.3 |1980.2  |1939.9|
|blur7_horz16_sse2|2645.1 |2659.1    |2615.1 |
|mul_bitmaps_sse2|71.6 |72.6 |70.3 |
|fill_halfplane_tile32_sse2|249.0  |248.2  |239.4 |
|blur8_vert16_sse2|4361.6  |4376.3   |4363.0 |
|shrink_vert16_sse2|335.4  |344.0     |339.4|
|blur8_horz16_sse2|2793.7  |2805.5    |2767.1 |
|be_blur_sse2|152.7   |157.5|149.7|
|stripe_pack16_sse2|125.1 |100.8   | 123.4|
|shrink_horz16_sse2|581.0      |586.9     |579.6 |
|expand_vert16_sse2|379.5  |382.1   |383.0 |
|blur4_vert16_sse2|1741.2 |1752.6  |1738.2|
|fill_solid_tile32_sse2|15.7  |16.1   |15.7 |
|expand_horz16_sse2|551.7  |557.0    |551.5 |
|blur6_vert16_sse2|2904.5 |2921.3  |2894.6 |
|blur4_horz16_sse2|1267.4 |1274.0     |1258.2|
|blur6_horz16_sse2|2277.5  |2292.6 |2251.5 |
|fill_halfplane_tile16_sse2|68.9  |67.6  |67.2 |
|fill_generic_tile32_sse2|271.8 |277.7 |276.4 |
|blur7_vert16_sse2|3599.8    |3618.9  |3600.7 |
|merge_tile32_sse2|43.8 |44.3 |44.0 |
| | | | |
|**SUM**|**29063.9**|**29217.1**|**28941.4**|

There being a difference at all suggests checkasm’s C code does indeed end up using AVX instructions close to the ASM routines *(and any overhead zeroing happening in the part which is still timed)*, but ideally I would have liked to see benchmarks of full file renders with forced SSE2 ASM routines, not just individual routines, to get a full picture and results robust against where the overhead occurs. Still, these results at least suggest there’s no regression. *(Individual results sometimes lag behind the unpatched `-mno-vzeroupper` test, but the  difference is likely not statistically significant)*

> I thought autotools was close to obsolete for libass.

The status of both build systems is clearly document in our README and the release announcement introducing meson. For most platforms autotools will remain the _preferred_ build system for the foreseeable future. You also ran into meson insufficiencies and should’ve seen warnings being logged about it over in your other PR.

TheOneric commented:
Your meson changes only enabled the conversion on x86_64 but not 32-bit x86, while autoconf enabled it on all x86 platforms.

The [configure check in dav1d](https://code.videolan.org/videolan/dav1d/-/commit/64c65de3faf1fa5e0272858203a9590f4ca31927) from which you appear to have copied the comment verbatim enables is for all x86 platforms.
I pushed an update resolving this plus style issues in the autoconf part.

Andarwinux commented:
Thanks! I did indeed copy it from dav1d. Do you think we should also add a -mavx CI?

TheOneric commented:
Since this is a x86asm feature and the CI doesn't actually verify SSE2 instructions got converted to AVX, I don’t think they add anything.

Given there were no more comments, I merged it now. Thanks!
